Remove use of np.matrix in coordinates, version 2

mhvk opened the pull request:
This is the second of two possible ways to remove the use of np.matrix in coordinates (see #5088).

In this PR, the matrices are replaced with normal arrays, and a new `matmul` function is defined that either just uses the numpy function of that name, or, for numpy <= 1.10, the equivalent `einsum`. With this implementation, in all the transformations, multiplication has to be replaced with `matmul`, so this affects a lot more files. To get general broadcasting to work, it will be necessary to also replace any `.T` on the rotation matrices with `.swapaxes(-2, -1)`.

mhvk commented:
p.s. If we go this route, `matmul` should probably be defined in `utils.compat.numpy` rather than in `angles`, but this can obviously be done later.

astrofrog commented on astropy/coordinates/builtin_frames/galactic_transforms.py:
```diff
@@ -23,7 +23,7 @@ def fk5_to_gal(fk5coord, galframe):
     mat3 = rotation_matrix(Galactic._ngp_J2000.ra.degree, 'z')
 
-    return mat1 * mat2 * mat3 * pmat
+    return matmul(matmul(matmul(mat1, mat2), mat3), pmat)
```
I wonder if we should have a standard way of indicating 'review this when we only support 3.5 and higher', or 'review this when we only support numpy 1.11 or higher', etc.?

bsipocz replied:
What about having a very distinct color and naming for a label? Something like "revise after: np 1.11"?

astrofrog replied:
I think this should be something in the code though, not just labels. But yes, a syntax like:

``# REVISE PY35: ...``

could work

mhvk replied:
It indeed needs to be something within a comment, since one cannot have `@` as part of the code, even if protected by an `if` statement.

I do think, though, that the code will look clean enough if I define `matrix_product` (which repeatedly multiplies matrices together) and perhaps also a `matrix_transpose`.

MSeifert04 replied:
You could wrap it in an `if py35: eval("a @ b")` but granted that's not a good choice. But the `@` operator is really bad for backwards-compatibility because implementing it means that older python are not just not supported anymore but it will be impossible to use it because of the SyntaxErrors. So maybe using `eval()` until 2025 would be the best way if astropy doesn't want to **exclude** any older python version.

mhvk commented:
Since the move to `matmul` seemed to be preferred, I went ahead and simplified expressions here by using `matrix_product` (which multiplies its arguments in sequence using `reduce`). I also made a `matrix_transpose` to have at least a hope multidimensional matrices will work (though I haven't added any test cases yet; PRs to my PR more than welcome...).

p.s. The only place where I did not yet remove `.T` is in `ecliptic_transforms`, since for that I feel #4293 should be resolved first.

astrofrog commented:
@mhvk - have you done any kind of performance benchmarking to see how things change with this PR?

mhvk commented:
Not beyond checking that `np.einsum(...)` is faster than multiplying to `np.matrix` instances.  I'll need to remind myself how to do asv...

mhvk commented:
@eteq, @StuartLittlefair - would you be able to review this PR to remove the use of `np.matrix`? It passes all tests, and I think it is an improvement in itself. Next up would be to add full tests for broadcasting of coordinates and times, etc., but my own sense is that that is better done as a separate PR, since this one is already quite large.

StuartLittlefair commented on astropy/coordinates/angles.py:
```diff
@@ -668,23 +667,20 @@ def __array_wrap__(self, obj, context=None):
+        Either ``'x'``, ``'y'``, ``'z'``, or a (x,y,z) specifying the axis to
+        rotate about. If ``'x'``, ``'y'``, or ``'z'``, the rotation sense is
+        counterclockwise looking down the + axis (e.g. positive rotations obey
+        left-hand-rule).  If given as an array, the last dimension should be 3.
```
Is it a good idea to also allow an array of ```str```s? For example ['x', 'y', 'z']?

mhvk replied:
'x', 'y', and 'z' are mostly meant as short-cuts to speed up those common cases; the speed-up would certainly be lost for multiple values, and since this is used mostly internally anyway, I'd tend to leave it alone.

StuartLittlefair commented on astropy/coordinates/builtin_frames/fk4_fk5_transforms.py:
```diff
@@ -49,19 +52,19 @@ def fk4_no_e_to_fk5(fk4noecoord, fk5frame):
     pmat2 = fk5frame._precession_matrix(EQUINOX_J2000, fk5frame.equinox)
 
-    return pmat2 * B * pmat1
+    return matrix_product(pmat2, B, pmat1)
```
I think this notation is very clean.

mhvk replied:
Yes, I was quite happy with `matrix_product` too. I'll admit to being a bit less happy with `matrix_transpose`, but at least it isn't needed all that often...

StuartLittlefair commented on astropy/coordinates/builtin_frames/utils.py:
```diff
@@ -34,29 +36,42 @@
+    """Matrix multiply all arguments together.
+
+    Arguments should have dimension 2 or larger, with larger dimensional
+    objects being interpreted as stacks of matrices.
```
Perhaps take a prompt from the ```numpy.matmul``` documentation and change this to "stacks of matrices residing in the last two indexes"?

StuartLittlefair commented on astropy/coordinates/builtin_frames/utils.py:
```diff
@@ -34,29 +36,42 @@
 def cartrepr_from_matmul(pmat, coo, transpose=False):
     """
     Note that pmat should be an ndarray, *not* a matrix.
     """
```
Should we include in the docstring here the way pmat is treated if it is ```N``` dimensional and ```N > 2```? I'm thinking of future ```coordinates``` developers trying to understand what shape their arrays should be...

mhvk replied:
OK, good idea. I updated the docstring.

StuartLittlefair commented:
@mhvk this is a 👍 from me.

mhvk commented:
@eteq - do you want to review this, or are you happy with @StuartLittlefair's review?

@astrofrog - on possible speed regressions: I *think* this should only lead to improvements, but still haven't figured out how to easily run asv. However, a quick test seems promising:
```
from astropy.coordinates import SkyCoord, Galactic
import astropy.units as u
sc = SkyCoord(np.arange(100.)*u.hourangle, np.zeros(100)*u.deg)
%timeit sc.transform_to(Galactic)
# Debian [1.1.2]: 10 loops, best of 3: 16.5 ms per loop
# This PR:       100 loops, best of 3: 14.3 ms per loop
```

mhvk commented:
@eteq - with @StuartLittlefair's review, I'm confident enough to merge this removal of the use of `np.matrix` in coordinates, but it might still be good if you could have a look as well. If not, one question: do you think this needs an entry in the API change section for 1.3? Obviously, it only changes internal routines, so strictly it is not necessary, but it might still be good to warn that `rotation_matrix` now returns `ndarray` just in case anybody has been using the routine for their own coordinate system transforms.

eteq commented:
I definitely plan to have a look at this, @mhvk.  It might take me a few more days but I'll try to bump it up in the queue seeing as how you think it's pretty much there.

One possible logistical complication: I think I've worked out a fix to #5133 that would make be made easier by use of this machinery. But ideally we would fix that in 1.2.x and this PR is certainly 1.3.x.  I might just try for a "quick-and-dirty" fix that would work in 1.2.x, but in that case we'd want to get that in *before* this.

mhvk commented:
@eteq - sounds good. I'm off on holidays at the end of the week, but happy to rebase beforehand with the fix for #5133 beforehand.

StuartLittlefair commented:
I've just discovered the ```transform``` method of ```CartesianRepresentation```.

Given the changes here, is this PR a sensible place to update that method to use the code contained within this PR?

mhvk commented:
That particular method cannot use `matmul` since one of the arguments is possibly a list of vectors. But it should use the revised `cartrepr_from_matmul` to deal properly with possible stacks of matrices (and avoid needless copying of data). I'll do that.

mhvk commented:
actually... more logical would be to adjust the method and use that in `cartrepr_from_matmul`...

mhvk commented:
OK, so I moved the code to `Cartesian.transform` and then realised `cartrepr_from_matmul` essentially had become a pass-through function, so I thought I might as well remove it altogether...

@eteq - if you feel this makes the PR expand too much beyond its original goal, I can revert the last two commits.

mhvk commented:
fixes #4878.
(the last two PRs fix broadcasting in `CartesianRepresentation.transform`)

eteq commented on astropy/coordinates/angles.py:
```diff
@@ -704,35 +700,37 @@ def rotation_matrix(angle, axis='z', unit=None):
-                          (0, -s, c)))
-    else:
+    try:
+        i = 'xyz'.index(axis)
```
Sneaky! (no action required - this is just positive feedback ;)

eteq commented on astropy/coordinates/builtin_frames/utils.py:
```diff
@@ -34,29 +35,18 @@
+    """Matrix multiply all arguments together.
+
+    Arguments should have dimension 2 or larger. Larger dimensional objects
+    are interpreted as stacks of matrices residing in the last two dimensions.
```
I think this docstring needs a bit more about *why* this function exists (which if I'm understanding right is being there mainly to ensure a consistent convention for dimensionality and so on).

mhvk replied:
OK, makes sense. Now done.

eteq commented on astropy/coordinates/builtin_frames/utils.py:
```diff
@@ -34,29 +35,18 @@
+
+
+def matrix_transpose(m):
+    """Transpose a matrix or stack of matrices by swapping the last two axes."""
```
similar to the above, it would be useful for this dosctring to justify why it exists  - someone not looking closely might say "why not just do ``M.T``?"

mhvk replied:
Also done. Seeing it again, I'm wondering slightly about just doing this explicitly... But maybe readability really wins here.  It will certainly not matter much for speed.

eteq commented on setup.cfg:
```diff
@@ -9,7 +9,7 @@ show-response = 1
 [pytest]
 minversion = 2.3.3
-norecursedirs = ".tox" "build" "docs[\/]_build" "astropy[\/]extern" "astropy[\/]utils[\/]compat"
+norecursedirs = ".tox" "build" "docs[\/]_build" "astropy[\/]extern" "astropy[\/]utils[\/]compat[\/]futures"
```
Can you clarify the point of this, @mhvk ? It's not clear to me how (if at all) it relates to the rest of this PR

mhvk replied:
This ensures that tests in `utils.compat.numpy` get run (by not excluding them any more). I checked that only `futures` had tests that do not work.

eteq commented:
@mhvk - sorry it's taken me ages to get to this, but I think I'm almost happy - I left one final question above, and one more thing here: this needs a changelog entry.  

This leads to a particular question, though: the changlog entry will need to note that rotation_matrix now returns an array rather than a matrix.  And that's actually an API change, right?  I wonder if that means we should instead change all uses of `rotation_matrix` to `rotation_matrix_array` or similar, and live `rotation_matrix` as a thin wrapper that just yields the (broken) numpy matrix class? 

And depending on how that goes, I think the sag example (http://docs.astropy.org/en/stable/generated/examples/coordinates/plot_sgr-coordinate-frame.html) has to be updated to reflect the new situation (looks like it should be a one-liner though)

mhvk commented:
I don't think `rotation_matrix` was ever exposed (it is not in `angles.__all__`), and hence assumed it was not part of the public API. Am I wrong?  I can certainly change it, though I think in that case we should add a deprecation warning on `rotation_matrix`, so we can eventually remove it.

mhvk commented:
@eteq - I also had not added a changelog entry because I assumed `rotation_matrix` was only used internally.  I'll do so once we've decided on how to treat it.

mhvk commented:
Oops, I should have read your note better: since there is an example of how to create a new frame that uses `rotation_matrix`, I think you are right we are stuck with supporting it, at least for a while. Should I still add a deprecation warning to it?

adrn commented:
@mhvk Yea, I know I have used it in some of my own code and it'd be useful to know where -- I think adding a deprecation warning is the way to go

eteq commented:
:+1: to a deprecation warning.  So maybe change the current `rotation_matrix` to `rotation_matrix_array` (or `rotation_array`)?  

And then also I guess it should be imported into `astropy.coordinates` so that it's actually documented as real public API.  I think the original concern was that it might overlap with some work in `modeling`, but it's pretty clear from @adrn's example that there are use cases where you definitely want to use it in creating a custom coordinate.  And along the same lines, I guess `matrix_transpose` and `matrix_product` should be made public. Another perturbation on that might be to document `coordinates.builtin_frames.utils` separately with a note at the top along the lines of "these are primarily for internal use in custom coordinates"?  Although maybe it's not *that* important to make a distinction.

mhvk commented:
@eteq, @adrn - I pushed a version where `angles.rotation_matrix` and `angles.angle_axis` are deprecated in favour of new routines in a new submodule `matrix_utilities`. I also moved `matrix_product` and `matrix_transpose` there. I have not yet imported these in `coordinates` itself, as I thought that might be a somewhat separate discussion.

Note that I just changed the location rather than the name of `rotation_matrix`, as I think the name is just too clear to let it go! Is a different location good enough? If not, I guess one could add `_3d`, but since this wasn't much in the public eye, I thought my present solution was OK.

mhvk commented:
@eteq, @adrn - could you check you're happy with the deprecation warning, and with how I moved routines to a new `matrix_utilities` file?  I'd like to merge this, as it is holding up further work on shape changes, etc.

adrn commented:
This looks good to me -- thanks @mhvk !

eteq commented:
Looks good to me too except for one super minor thing: I think this should go in the "other" section of the changelog, as it isn't really a bug fix (but isn't exactly an API change due to it not explicitly being public API).  Will just do that in a manual merge though so as not to delay things further.  

Thanks for all the work on this, @mhvk!

eteq commented:
and just for the record, if anyone else is looking at this and wondering, the travis failures were a separate problem, reported in #5198 and fixed in #5203

eteq commented:
@mhvk - didn't think this needed delaying the merge, but I see you modified the Sgr example to use the new matrices but left in this:
```
SGR_MATRIX = np.array(B.dot(C).dot(D))
```

That could be changed to
```
SGR_MATRIX = matrix_product(B, C, D)
```
right?

mhvk commented:
Yes, sorry I missed that...  Will look at #5210 tomorrow.
